esmvaltool/diag_scripts/cos22esd/hotspot_plotter.py

```diff
@@ -0,0 +1,647 @@
+
+        N indicates the number of models included in the ensemble mean.
+        """
+        top, bottom, left, right = 0.75, 0.2, 0.02, 0.99
+        wspace, hspace = 0.005, 0.005
+        sorted_keys = [f"{period}_{season}_{variable}_{project}_{scenario}"
+                       for variable in self.variables
+                       for period in self.cfg["future_periods"]
+                       for project in self.projects for season in self.seasons]
+        sorted_keys = [
+            sorted_keys[:len(sorted_keys) // 2],
+            sorted_keys[len(sorted_keys) // 2:]]
+        ancestor_files_var = [
+            [ancestor_file for ancestor_file in ancestor_files_var
+             if f"/{var}_" in ancestor_file] for var in self.variables]
+        for ancestor_files, keys, variable in zip(ancestor_files_var,
+                                                  sorted_keys,
+                                                  self.variables):
+            fig = plt.figure(figsize=(14.4, 3.4),
+                             constrained_layout=True, dpi=300)
+            plt.gcf().subplots_adjust()
+            proj, projection, path_ext, plotextend = self.define_projection(
+                self.cfg["region"])
+            gspec = GridSpec(
+                len(self.cfg["future_periods"]),
+                len(self.seasons) * 2,
+                figure=fig,
+                hspace=hspace,
+                wspace=wspace,
+                top=top,
+                bottom=bottom,
+                left=left,
+                right=right,
+            )
+
+            # bound colorbar to abs(max) value on the map
+            # or use input bounds
+            if variable == "tas":
+                if tas_bound:
+                    bound_limit = tas_bound
+                else:
+                    bound_limit = self.find_abs_bound_range(results_dict, keys)
+                cmap = plt.cm.RdBu_r
+            else:
+                if pr_bound:
+                    bound_limit = pr_bound
+                else:
+                    bound_limit = self.find_abs_bound_range(
+                        results_dict, keys, avg_over=25)
+                cmap = plt.cm.BrBG
+            bounds = np.linspace(-1 * bound_limit, bound_limit, 11)
+
+            # plot each pannel
+            for i, key in enumerate(keys):
+                if i < 6:
+                    axes = fig.add_subplot(gspec[0, i], projection=proj)
+                    plt.title(f"{key.split('_')[1].upper()}")
+                else:
+                    axes = fig.add_subplot(gspec[1, i - 6], projection=proj)
+
+                if projection == "LambertConformal":
+                    proj_to_data = (
+                        ccrs.PlateCarree()._as_mpl_transform(axes) - axes.transData
+                    )
+                    rect_in_target = proj_to_data.transform_path(path_ext)
+                    axes.set_boundary(rect_in_target, use_as_clip_path=True)
```

Comment:
I am not entirely sure what is this doing, but is there any alternative to using a private function? It looks like [`set_boundary` has a `transform` argument](https://scitools.org.uk/cartopy/docs/v0.13/_modules/cartopy/mpl/geoaxes.html#GeoAxes.set_boundary) that calls this same function, is it possible to use that instead?

Reply:
I think @sloosvel is bang on here. See also [this cartopy explanation](https://scitools.org.uk/cartopy/docs/latest/tutorials/understanding_transform.html). To spell it out, here it should be enough to pass the rect in source coordinates along with the right crs as transform:
```suggestion
                    axes.set_boundary(rect_in_source, transform=ccrs.PlateCarree(), use_as_clip_path=True)
```

Reply:
Thanks for the suggestion @zklaus. I'm not entirely sure what the `rect_in_source` variable represents, though.

Reply:
Sorry, I should have been more explicit. That is meant to be the path in source coordinates, i.e. plate carree. You can try with just the four corners of the rectangle. You might then run into the interpolation problem, though [this example](https://scitools.org.uk/cartopy/docs/latest/matplotlib/intro.html#adding-data-to-the-map) seems to suggest that you might be fine. If not, you still may need to do the interpolation that you already have in [define_projection](https://github.com/ESMValGroup/ESMValTool/blob/89a2d9b1c9dc5b8e979a1fbadc42499b3ddba3ef/esmvaltool/diag_scripts/cos22esd/hotspot_plotter.py#L667), but I would first try straight up the uninterpolated path in lon-lat coordinates.

Reply:
You are right, directly using the rectangle bounds and then transforming it inside the set_boundary works fine. No need to do it before.